salt/modules/freezer.py

```diff
@@ -0,0 +1,291 @@
+
+    As this module is build on top of the pkg module, the user can
+    send extra attributes to the underlining pkg module via kwargs.
+
+    name
+        Name of the frozen state. Optional.
+
+    force
+        If true, overwrite the state. Optional.
+
+    kwargs
+        Extra parameters send to the underlining pkg module. Optional.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' freezer.freeze
+        salt '*' freezer.freeze force=True root=/chroot
+
+    '''
+    states_path = _states_path()
+    ret = __salt__['cmd.retcode'](['mkdir', '-p', states_path])
+    if ret:
+        msg = 'Error when trying to create the freezer ' \
+            'storage: {}'.format(states_path)
+        log.error(msg)
+        raise CommandExecutionError(msg)
+
+    if isfrozen(name) and not force:
+        raise CommandExecutionError('The state is already present. Use '
+                                    'force parameter to overwrite.')
+    safe_kwargs = {k: v for k, v in kwargs.items() if not k.startswith('__')}
+    pkgs = __salt__['pkg.list_pkgs'](**safe_kwargs)
+    repos = __salt__['pkg.list_repos'](**safe_kwargs)
+    for name, content in zip(_paths(name), (pkgs, repos)):
+        with fopen(name, 'w') as fp:
+            json.dump(content, fp)
+    return True
+
+
+def restore(name=None, **kwargs):
+    '''
+    Make sure that the system contains the packages and repos from a
+    frozen state.
+
+    Read the list of packages and repositories from the freeze file,
+    and compare it with the current list of packages and repos. If
+    there is any difference, all the missing packages are repos will
+    be installed, and all the extra packages and repos will be
+    removed.
+
+    As this module is build on top of the pkg module, the user can
+    send extra attributes to the underlining pkg module via kwargs.
+
+    name
+        Name of the frozen state. Optional.
+
+    kwargs
+        Extra parameters send to the underlining pkg module. Optional.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' freezer.restore
+        salt '*' freezer.restore root=/chroot
+
+    '''
+    if not isfrozen(name):
+        raise CommandExecutionError('Frozen state not found.')
+
+    frozen_pkgs = {}
+    frozen_repos = {}
+    for name, content in zip(_paths(name), (frozen_pkgs, frozen_repos)):
+        with fopen(name) as fp:
+            content.update(json.load(fp))
+
+    # The ordering of removing or adding packages and repos can be
+    # relevant, as maybe some missing package comes from a repo that
+    # is also missing, so it cannot be installed. But can also happend
+    # that a missing package comes from a repo that is present, but
+    # will be removed.
+    #
+    # So the proposed order is;
+    #   - Add missing repos
+    #   - Add missing packages
+    #   - Remove extra packages
+    #   - Remove extra repos
+
+    safe_kwargs = {k: v for k, v in kwargs.items() if not k.startswith('__')}
```

Comment:
We have [utility](https://github.com/saltstack/salt/blob/develop/salt/utils/args.py#L30) exactly for that. Please use it instead.

Reply:
indeed